src/platformio.ts

```diff
@@ -85,21 +98,124 @@ export class PlatformIO {
+                    this.showErrorToAddIncludePath();
+                }
+
                 return;
             }
-            vscode.window.showInformationMessage("Add Include Path to c_cpp_properties.json successfully.");
+
+            if (!silent) {
+                vscode.window.showInformationMessage("Add Include Path to c_cpp_properties.json successfully.");
+            }
         } catch (ex) {
-            this.showErrorToAddIncludePath();
+            if (!silent) {
+                this.showErrorToAddIncludePath();
+            }
         } finally {
-            AppInsightsClient.sendEvent("addIncludePath");
+            if (!silent) {
+                AppInsightsClient.sendEvent("addIncludePath");
+            }
         }
     }
 
     public onDidCloseTerminal(closedTerminal: vscode.Terminal): void {
         this._pioTerminal.onDidCloseTerminal(closedTerminal);
     }
 
+    private configureAutoUpdateIncludes(): void {
+        const cppExtension = vscode.extensions.getExtension("ms-vscode.cpptools");
+        if (cppExtension) {
+            const autoUpdateIncludes = vscode.workspace.getConfiguration("platformio").get<boolean>("autoUpdateIncludes");
+            if (autoUpdateIncludes && !this._autoUpdateIncludesEnabled) {
+                this.buildFolderList();
+                const configFilePath = path.join(vscode.workspace.rootPath, ".vscode", "c_cpp_properties.json");
+                if (!Utility.fileExistsSync(configFilePath)) {
+                    this.addIncludePath(true);
+                }
+
+                // Watch .piolibdeps and lib folders
+                const fileSystemWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/{.piolibdeps,lib}/**");
+                this._autoUpdateIncludesDisposables.push(fileSystemWatcher);
+                fileSystemWatcher.onDidCreate((e: vscode.Uri) => this.fileSystemDidCreate(e), this, this._autoUpdateIncludesDisposables);
+                fileSystemWatcher.onDidDelete((e: vscode.Uri) => this.fileSystemDidDelete(e), this, this._autoUpdateIncludesDisposables);
+
+                // Watch platformio.ini file
+                const iniFileWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/platformio.ini");
+                this._autoUpdateIncludesDisposables.push(iniFileWatcher);
+                iniFileWatcher.onDidChange((e: vscode.Uri) => this.addIncludePath(true), this, this._autoUpdateIncludesDisposables);
+                iniFileWatcher.onDidCreate((e: vscode.Uri) => this.addIncludePath(true), this, this._autoUpdateIncludesDisposables);
+            }
+
+            if (!autoUpdateIncludes && this._autoUpdateIncludesEnabled) {
+                this._folderList = undefined;
+                this.disposeAutoUpdateIncludes();
+            }
+
+            this._autoUpdateIncludesEnabled = autoUpdateIncludes;
+        }
+    }
+
+    private fileSystemDidCreate(e: vscode.Uri): void {

```

Comment:
It has become more complicated than I like - as you can probably guess from what I wrote before, the ```FilesystemWatcher``` is not my best friend.

There are two instances of ```FilesystemWatcher```:

```typescript
const fileSystemWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/{.piolibdeps,lib}/**");

...

const iniFileWatcher: vscode.FileSystemWatcher = vscode.workspace.createFileSystemWatcher("**/platformio.ini");

```

The first ```FilesystemWatcher``` triggers when something is created or deleted in the .piolibdeps and lib folders. 

When something is created, ```fileSystemDidCreate(e: vscode.Uri)``` will be called and that is the method that have caused me most trouble.

In my first attempt, I tried to be clever and it looked like this:

```typescript
private fileSystemDidCreate(e: vscode.Uri): void {
    let parent = this.getParentFolder(e.path);
    if (!this._folderList.has(parent)) {
        this._folderList.add(parent);
        this.debounceChanges();
    }
}
```

My idea was that if I just look at the parent folder, then I could build the list of folders when something got added to the folder. I would not get the an empty folder that stayed empty that way, but we don't need empty folders for includes.

I liked that solution because it was fast without any filesystem calls (I stress tested with a folder I have with 100k mostly small files in 20k folders), but then I tried to install library 303 with PIO and got events like this:

```
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\README.md
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\RestClient.h
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\RestClient.cpp
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\library.json
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\LICENSE.md
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\library.properties
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\examples
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\examples\simple_POST
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\examples\simple_POST\simple_POST.ino
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\examples\full_test_suite
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\examples\full_test_suite\full_test_suite.ino
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\examples\simple_GET
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\examples\simple_GET\simple_GET.ino
Create: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package\.library.json
Create: d:\Projects\ESP8266Test\.piolibdeps\ESP8266RestClient_ID303
Delete: d:\Projects\ESP8266Test\.piolibdeps\_tmp_installing-qkf_d2-package
```

PIO unpacks everything in a temporary folder and finally renames it. That means the ESP8266RestClient_ID303 folder dosen't get added to my list of folders and I will not see it as a folder if it gets deleted.

Then I tried a version where I check every path to see if it is a folder:

```typescript
private fileSystemDidCreate(e: vscode.Uri): void {
    try {
        if (fs.statSync(e.fsPath).isDirectory()) {
            if (!this._folderList.has(e.path)) {
                this._folderList.add(e.path);
                this.debounceChanges();
            }
        }
    } catch (error) {
        // Ignore errors, the file or folder was probably temporary and was
        // renamed or deleted before we could check if it was a folder.
    }
}
```

It slowed down my stress test with about 20%, which is not as bad as I feared, but when I counted the number of times it was triggered, I see that it missed a few changes. I don't undertsand enough about how the internals of Node's filesystem code works and it could also be something in VS Code, so I just ended up with combining both and see that my stress test ended up with the correct number of folders. It could be related to this issue: https://github.com/Microsoft/vscode/issues/26578

If we agree that we don't care about deleted folders, then I can go back to the first version and get rid of the code than handles deletes.

The second ```FilesystemWatcher``` triggers when the platformio.ini file is changed or created and calls ```addIncludePath(true)``` immediately.